tutor/commands/upgrade/compose.py

```diff
@@ -166,16 +167,53 @@ def upgrade_from_olive(context: click.Context, config: Config) -> None:
 
     click.echo(fmt.title(f"Upgrading MySQL to {intermediate_mysql_docker_image}"))
 
-    # Revert the MySQL image to build the data dictionary on v8.1
-    mysql_docker_image = config["DOCKER_IMAGE_MYSQL"]
-    config["DOCKER_IMAGE_MYSQL"] = intermediate_mysql_docker_image
+    # We start up a mysql-8.1 container to build data dictionary to preserve
+    # the upgrade order of 5.7 -> 8.1 -> 8.4
+    # Use the mysql-8.1 context so that we can clear these filters later on
+    with hooks.Contexts.app("mysql-8.1").enter():
+        hooks.Filters.ENV_PATCHES.add_item(
+            (
+                "local-docker-compose-services",
+                """
+mysql-8.1:
+  extends: mysql
+  image: docker.io/mysql:8.1.0
+  command: >
+    mysqld
+    --character-set-server=utf8mb3
+    --collation-server=utf8mb3_general_ci
+    --binlog-expire-logs-seconds=259200
+    """,
+            )
+        )
+        hooks.Filters.ENV_PATCHES.add_item(
+            (
+                "local-docker-compose-jobs-services",
+                """
+mysql-8.1-job:
+  image: docker.io/mysql:8.1.0
+  depends_on: {{ [("mysql-8.1", RUN_MYSQL)]|list_if }}
+        """,
+            )
+        )
+        hooks.Filters.CONFIG_DEFAULTS.add_item(("MYSQL_HOST", "mysql-8.1"))
+
+        hooks.Filters.CLI_DO_INIT_TASKS.add_item(
+            ("mysql-8.1", tutor_env.read_core_template_file("jobs", "init", "mysql.sh"))
+        )
+
     tutor_env.save(context.obj.root, config)
+
     # Run the init command to make sure MySQL is ready for connections
-    context.invoke(jobs.initialise, limit="mysql")
+    context.invoke(jobs.initialise, limit="mysql-8.1")
+    context.invoke(compose.stop, services=["mysql-8.1"])
 
-    # Change the image back to v8.4
-    config["DOCKER_IMAGE_MYSQL"] = mysql_docker_image
+    # Clear the filters added for mysql-8.1 as we don't need them anymore
+    hooks.clear_all(context="app:mysql-8.1")
+
+    # Save environment and run init for mysql 8.4 to make sure MySQL is ready
     tutor_env.save(context.obj.root, config)
+    context.invoke(jobs.initialise, limit="mysql")
```

Comment:
Where is mysql version set to 8.4? Or is it 8.4 already before the code enters MySQL 8.1 context?

Reply:
It is already set prior to the context when the user installs tutor v18 onwards.